xds/internal/balancer/xdsrouting/matcher_header.go

```diff
@@ -0,0 +1,240 @@
+/*
+ *
+ * Copyright 2020 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package xdsrouting
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"google.golang.org/grpc/balancer"
+	"google.golang.org/grpc/metadata"
+)
+
+func mdValuesFromOutgoingCtx(ctx context.Context, key string) ([]string, bool) {
+	if ctx == nil {
+		return nil, false
+	}
+	md, ok := metadata.FromOutgoingContext(ctx)
+	if !ok {
+		return nil, false
+	}
+	vs, ok := md[key]
+	return vs, ok
+}
+
+type headerExactMatcher struct {
```

Comment:
Might be worth getting rid of the `header` prefix from these matcher names. All the matchers here are for headers, and in any case, the caller will be using the interface and not the exact types, so having the `header` prefix in the matcher names is not so much useful.